LiteDB/Engine/Disk/MemoryCache.cs

```diff
@@ -57,29 +57,50 @@ public PageBuffer GetReadablePage(long position, FileOrigin origin, Action<long,
         {
             // get dict key based on position/origin
             var key = this.GetReadableKey(position, origin);
-
-            // try get from _readble dict or create new
-            var page = _readable.GetOrAdd(key, (k) =>
+            do
             {
-                // get new page from _free pages (or extend)
-                var newPage = this.GetFreePage();
-
-                newPage.Position = position;
-                newPage.Origin = origin;
-
-                // load page content with disk stream
-                factory(position, newPage);
-
-                return newPage;
-            });
-
-            // update LRU
-            Interlocked.Exchange(ref page.Timestamp, DateTime.UtcNow.Ticks);
+            	//If key is not present, just create new page
+                if (!_readable.TryGetValue(key, out var page))
+                {
+                    var newPage = GetFreePage();
+                    newPage.Position = position;
+                    newPage.Origin = origin;
+                    newPage.ShareCounter = 1;
+                    Interlocked.Exchange(ref newPage.Timestamp, DateTime.UtcNow.Ticks);
 
-            // increment share counter
-            Interlocked.Increment(ref page.ShareCounter);
+                    if (_readable.TryAdd(key, newPage))
+                    {
+                        // load page content with disk stream
+                        factory(position, newPage);
+                        return newPage;
+                    }
 
-            return page;
+                    newPage.ShareCounter = 0; //Wasn't able to add it? discard and try again
+                    _free.Enqueue(newPage);
```

Comment:
before return to `_free` it should be cleaned up from old data (check how Extend does with released pages). Method may be extracted for that.